### PR TITLE
Disambiguate expansions relative to the creating DepNode

### PR DESCRIPTION
**Pull request by cjgillot**
Expansion data requires to be disambiguated in case of colliding spans. The current disambiguation mechanism was merely looking at hash collisions using a global disambiguation counter.

This global counter created an untracked data dependency between unrelated queries. For instance here: https://github.com/rust-lang/rust/issues/110457#issuecomment-1555008573

This PR replaces the global disambiguation map by a per-DepNode map. This requires threading the currently evaluated DepNode through the implicit context. We use a dummy in the non-incremental case to simplify the logic.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/context/tls.rs [hunk 1]
```diff
@@ -31,6 +31,10 @@ pub struct ImplicitCtxt<'a, 'tcx> {
     /// Used to prevent queries from calling too deeply.
     pub query_depth: usize,
 
+    /// The DepNode of the query being executed. This is updated by the dep-graph. Thisis used to
```
Comment (Nilstrieb) [inline compiler/rustc_middle/src/ty/context/tls.rs:34]:
```suggestion
    /// The DepNode of the query being executed. This is updated by the dep-graph. This is used to
```

**Review by oli-obk — APPROVED** [review]
this makes sense to me and the implementation does what you describe, but I'm far from fully grokking all the implications. I still think we should land it, unless you want to have someone else review it additionally.

**Comment by cjgillot** [comment]
As this touches the expansion infra, cc @petrochenkov

**Comment by petrochenkov** [comment]
I'm sufficiently sure that neither AST lowering nor *especially* MIR inlining should reuse the macro expansion machinery, just never got to investigating how to get rid of it.
I was never pinged on the relevant MIR change, so encountering `ExpnKind::Inlined` later was quite a WTF moment.

**Comment by petrochenkov** [comment]
I don't really understand how dep nodes work in this context, but assume the issue in #110457 happens because expansions are created "after incremental is enabled", and earlier expansions don't have such issues because they are still at the "redo everything and hash all the results" stage?

File: compiler/rustc_span/src/hygiene.rs [hunk 2]
```diff
@@ -350,14 +352,6 @@ impl ExpnId {
     }
 }
 
-/// This struct is meant to be a surrogate for the actual `DepNode` in rustc_middle.
-/// Both types should be kept in sync.
-#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
-pub struct FakeDepNode {
-    pub kind: u16,
-    pub hash: PackedFingerprint,
-}
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs]: There are some back and forth changes in the commits, probably makes sense to cleanup the history.

File: compiler/rustc_span/src/hygiene.rs [hunk 3]
```diff
@@ -1484,44 +1471,38 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
 #[instrument(level = "trace", skip(ctx), ret)]
 fn update_disambiguator(
     expn_data: &mut ExpnData,
-    dep_node: FakeDepNode,
+    hash_extra: impl Hash + Copy + fmt::Debug,
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1474]:
```suggestion
    hash_extra: impl Hash + Copy
```
Unused bound (in a couple of other places too).

File: compiler/rustc_query_system/src/dep_graph/mod.rs [hunk 4]
```diff
@@ -137,13 +140,27 @@ impl FingerprintStyle {
...
+    },
     Anonymous,
     Untracked,
 }
 
+impl<K> CurrentDepNode<K> {
+    pub fn regular(dep_node: DepNode<K>) -> CurrentDepNode<K> {
+        CurrentDepNode::Regular {
+            dep_node,
+            expn_disambiguators: Lrc::new(Lock::new(UnhashMap::default())),
```
Comment (petrochenkov) [inline compiler/rustc_query_system/src/dep_graph/mod.rs:159]:
```suggestion
            expn_disambiguators: Default::default(),
```
Fingers crossed.

**Comment by oli-obk** [comment]
> I'm sufficiently sure that neither AST lowering nor _especially_ MIR inlining should reuse the macro expansion machinery, just never got to investigating how to get rid of it.

Easy, we can just nuke it without replacement. The use case for it has since been refactored away, so at best it would cause a minor mir opt change

**Comment by cjgillot** [comment]
> I don't really understand how dep nodes work in this context, but assume the issue in #110457 happens because expansions are created "after incremental is enabled", and earlier expansions don't have such issues because they are still at the "redo everything and hash all the results" stage?

Exactly.

The ICE https://github.com/rust-lang/rust/issues/110457 itself has been fixed by https://github.com/rust-lang/rust/pull/111952. The theoretical ICE still exists, but I don't think there is a way to trigger it right now.

This global disambiguation is a footgun for incremental, so I'm still proposing to land this PR.

**Comment by petrochenkov** [comment]
The initial idea was to create expansion disambiguators per parent `DefPath` (https://github.com/rust-lang/rust/issues/49300#issuecomment-525531109).
During actual implementation @Aaron1011 found that to be too expensive and came up with the current solution.

This solution is pretty similar to per-defpath disambiguators, right? Maybe with a slightly different granularity?

**Comment by cjgillot** [comment]
Yes, it's quite close.
We may not achieve the same level of hash-stability as using a per-DefPath disambiguator: all expansions created using macro expansion correspond to the `resolver_for_lowering` DepNode.
OTOH, this scheme ensures that we don't have data flowing across query boundaries, which is important for incremental soundness.

As the DefPath approach was limited by perf:
@bors try @rust-timer queue

**Comment by Zoxc** [comment]
Could we limit the current disambiguation mechanism to just macro expansion say via. a scoped thread local?

I'd prefer to avoid extending the query system here, but I don't see any obviously wrong. It does seem this could use the `query` field of `ImplicitCtxt` instead of adding a new one, such that each query would have it's own counter.

**Comment by cjgillot** [comment]
This is not required any more, the bugs have been fixed independently. To be revisited if queries advance into the resolve.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/111950?
I assumed the issue existed for spans created during AST lowering too.

**Comment by cjgillot** [comment]
Both `ExpnKind::Inlined` (#111950) and `DesugaringKind::Replace` (#111952) have been removed, so we don't create new expansions after lowering.
We may have the issue with lowering if lowering stops being a monolithic query. For now, we should not reach it, as lowering gets reexecuted in each compilation.

**Pull request closed without merging**